suffuse/node_fuse.go

```diff
@@ -3,22 +3,92 @@ package suffuse
 import (
   "github.com/suffuse/go-suffuse/suffuse/xattr"
   "os"
+  "time"
   "golang.org/x/net/context"
   "bazil.org/fuse/fs"
   f "bazil.org/fuse"
 )
 
+// Access checks whether the calling context has permission for
+// the given operations on the receiver. If so, Access should
+// return nil. If not, Access should return EPERM.
+//
+// Note that this call affects the result of the access(2) system
+// call but not the open(2) system call. If Access is not
+// implemented, the Node behaves as if it always returns nil
+// (permission granted), relying on checks in Open instead.
 func (x *Elem) Access(ctx context.Context, req *f.AccessRequest) error {
   logD("Access", "path", x.Path)
+  // TODO - permissions check.
+  return nil
+}
+
+// Setattr sets the standard metadata for the receiver.
+//
+// Note, this is also used to communicate changes in the size of
+// the file. Not implementing Setattr causes writes to be unable
+// to grow the file (except with OpenDirectIO, which bypasses that
+// mechanism).
+//
+// req.Valid is a bitmask of what fields are actually being set.
+// For example, the method should not change the mode of the file
+// unless req.Valid.Mode() is true.
+func (x *Elem) Setattr(ctx context.Context, req *f.SetattrRequest, resp *f.SetattrResponse) error {
+  logD("Setattr", "path", x.Path, "req", *req)
+  // TODO - permissions check on ops which require it (see also Access.)
+  // TODO - accumulate and report errors somehow.
+
+  // There doesn't appear to be any way to set one or the other
+  // time, you always have to specify two values. So if both are
+  // valid according to fuse, we pass both, but if only one is,
+  // we stat the current values and pass what was already there
+  // for the not-being-set-right-now slot.
+  if req.Valid.Atime() || req.Valid.Mtime() {
+    var atime, mtime time.Time
+    if req.Valid.Atime() && req.Valid.Mtime() {
+      atime = req.Atime
+      mtime = req.Mtime
+    } else {
+      // one of them needs to be preserved
+      atime, mtime = x.Path.OsStatAtimeMtime()
+      if req.Valid.Atime() { atime = req.Atime }
+      if req.Valid.Mtime() { mtime = req.Mtime }
+    }
+    x.Path.OsChtimes(atime, mtime)
+  }
```

Comment:
If neither is valid then nobody is trying to change the times. Only the bits which match the Valid mask are relevant. 

If one is valid then somebody is trying to change the time. But the only way to change any time is to change both.